### DISCUSSION TITLE
Graph improvements; into the multichainverse

### DISCUSSION THREAD

**Discussion by bitbeckers** [hypercerts-org/hypercerts#1265] [Ideas - where HIPs are born]
A place to discuss what we can do to improve the Graph. Since we're pushing the SDKs and marketplace everywhere this week, I think we can work on improving the Graph starting next week. 

The most obvious issue is that we can't support x-chain instances since 
1) the SDK connects to a single Graph.
2) the SDK requires a user to connect to determine which Graph to show
3) graphIDs are in theory overlapping. This can be managed but is more clean to manager on the Graph itself.

Additionally we can implement IPFS in the Graph for faster loading of hypercerts data.

## The SDK connects to a single Graph
For 1, since we're using urql, we can add an exchange that splits routes: 
[example 1](https://github.com/urql-graphql/urql/discussions/1523)
[example 2](https://antoinelin.com/en/posts/how-to-use-multiple-graphql-endpoints-with-urql)

Then, we can let the client query endpoints by mapping over the supported chainIDs.

## The SDK requires a user to connect to determine which Graph to show
When we have that implemented, we can work on 2 by separating "test" and "production". This would allow a user to connect in readOnly mode for test or prod using:

`const client = new HypercertClient("test")` with a log warning that this is readonly

`const client = new HypercertClient(chain: { id: 11155111})` defaults to test but should block tx calls on non-11155111 hypercerts

To identify those hypercerts, we will update the hypercert IDs. 

## Graph IDs are overlapping
Currently the hypercerts ID is a concatenation of `{contractAddress}-{tokenID}`. We could update this to `{chainId}-{contractAddress}-{tokenID}`, with chainID being human readable integers (e.g. 11155111).

We can work on refactoring the IDs and connecting to multiple Graph instances at the same time. The are both requirements for "test" and "production" instances of the SDK.

@Jipperism could you provide some feedback?

**Comment by ryscheng** [comment]
Ya the GraphIDs being overlapping is a problem. I think prefixing with chainID is a reasonable way to disambiguate, I just hope the migration to that new format isn't too painful

As far as supporting multi-chain, what do you think about (1) just having multiple clients (one per chain) vs (2) supporting a single client connecting to multiple chains?

**Reply by bitbeckers** [reply]
The challenge is that a lot of Graph clients they only connect to a single Graph endpoint (found this out with another app) and the graphclient from Graph Protocol has their own issues. One option that I'm using with `urql` in another app is adding a middleware that can  handle the routing (similar to what the Graph Protocol client does with their resolvers):

```js
import {
  makeOperation,
  Operation,
  Exchange,
} from "@urql/core";
import { pipe, tap, map } from "wonka";

// https://github.com/urql-graphql/urql/discussions/1523

const graphUrls = {
  fux: "https://api.thegraph.com/subgraphs/name/bitbeckers/fux-sepolia",
  werk: "https://api.thegraph.com/subgraphs/name/bitbeckers/werk-sepolia",
};

// You can use options for inputs if you want to make this reusable for  the community
export const routeSplitterExchange =
  (): Exchange =>
  ({ forward }) => {
    const processIncomingOperation = (operation: Operation): Operation => {
      if (operation) {
        let url = operation.context.url;

        switch (operation.context.target) {
          case "fux":
            url = graphUrls.fux;
            console.log("Calling fux");
            break;
          case "werk":
            url = graphUrls.werk;
            console.log("Calling werk");
            break;
          default:
            url = graphUrls.fux;
            console.log("Calling default");
            break;
        }

        // your own condition
        return makeOperation(operation.kind, operation, {
          ...operation.context,
          url,
        });
      }

      return operation;
    };

    return (ops$) => {
      return pipe(ops$, map(processIncomingOperation), forward);
    };
  };
```

And, but this should we tested, we can take an array of the supported chains and map overthem with with same query and resolve those promisses. 

TLDR: I'm leaning towards 2 ;)

**Reply by ryscheng** [reply]
@holkeb this is partially a UX question as well.

Do you want to be able to see Hypercerts from any chain all at once?
Or is it okay if we only show the Hypercerts on the chain that your wallet is connected to?

The second is easier atm, but how soon (if at all), will we need the first?

**Reply by holkexyz** [reply]
I definitely want this at some point.

For now it would be ok to just show the hypercerts on the chain that you are connected to since most are on Optimism. (And gainforest could build just specifically for Celo).

However: Would someone who isn't connected at all, see no hypercerts?

**Reply by bitbeckers** [reply]
> However: Would someone who isn't connected at all, see no hypercerts?

Yes that's the crux. I've received multiple feedbacks on the fact that people think the app is broken because they're not connected. Or they're simply confused why it isn't loading.

**Reply by holkexyz** [reply]
And for people who are just interested, it would be great to see hypercerts. How complex is it do show hypercerts from any chain all at once? (complex... like in dev days or so)

**Reply by bitbeckers** [reply]
It's all the steps outlines above, where the Graph is the key because that provides the IDs. 

**Napkin calculation:**
Graph - 1 day
SDK + graph client in SDK - 1-2 days (incl. supporting test/production environment which enables this)
Plasmic - need to check but I believe it follows the IDs from the Graph

If we want to go this route I can map out the steps/features and get a better guestimate

**Reply by holkexyz** [reply]
Based on that guestimate, I think we should implement it

**Reply by bitbeckers** [reply]
I'll refine it more and if there's no big deviation we can kick that off (maybe next week)

**Reply by Jipperism** [reply]
Another thing to consider is the conditionally disabling/enabling of lots of buttons in the frontend, depending on which chain you are connected to, as reading for different chains is possible but write operations will only be available for the chain currently connected to. Then we would have to display the chain on hypercerts as well, to prevent confusion.

**Reply by bitbeckers** [reply]
@Jipperism could you help refine that on the FE/Plasmic level? Since these changes need to propagate in the frontend we also should consider a flow where we have a develop environment that breaks production. This bring us back to figuring out to to separate dev from prod in Plasmic.

**Reply by bitbeckers** [reply]
https://github.com/hypercerts-org/hypercerts/issues/1059

**Reply by Jipperism** [reply]
Went through the hypercerts.org frontend to get an overview of what we would need to update:

`/app/dashboard`
- 'claim all fractions' button should select only hypercerts on the current chain to claim. but to avoid confusion we would probably want to show allowlist entries from all chains. that would mean adding some sort of tag/banner to the list entries to display which chain they are on, and possibly make them greyed out so it's clear they won't be minted when pressing the button.

`/app/gallery-all`
- display a chip/banner on every tile to make it clear what chain the hypercert lives on.

`/app/create`
- display chain in minting confirmation. this should be obvious already, but as it will become more of a cross-chain users will probably be less aware of what chain they're on so it might be useful to add some sort of extra warning.

`/app/view?claimId=?`
- split, merge, transfer button should be disabled if current chain is not the same as the chain the hypercert lives on.

Then there is probably some stuff that needs to be updated in supabase, as there are a lot of id's stored there. We usually save the chain Id so probably just a simple scripts that loops and updates the id by concatenating those.

I can't access plasmic right now so not entirely sure how much of the data fetching would have to be updated, but to get everything working I would expect this to be at least a day of work.

**Reply by bitbeckers** [reply]
@Jipperism did you manage to get access to Plasmic? The biggest blocker to resolve I see right now is to have a clear plan on how to not break everything since Plasmic is the component that only has one instance.

**Comment by Jipperism** [comment]
+1 for adding IPFS integration to the graph. The current solution won't scale at all for any operations with regards to filtering/sorting, as you'd have to pull in metadata for all hypercerts separately.

**Reply by bitbeckers** [reply]
See #1269 for the issues that are starting to emerge with the gateways. As far as we know, the Graph endpoints don't have this issue. So instead of shooting for hail across three gateways, we're at the point of needing to provide metadata via a 'cache' that we provide. The Graph is the prime candidate here. 

Creating a separate issue for this, here's a good guide on the steps to take: https://blog.developerdao.com/how-to-properly-request-json-metadata-stored-in-ipfs-for-your-the-graph-subgraph

**Reply by bitbeckers** [reply]
#932
